### PR TITLE
C++: Add sscanf and fscanf models

### PR DESCRIPTION
**Pull request by MathiasVP**

### CODE REVIEW

File: cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+        "_snscanf_l", // _snscanf_l(src, length, format, locale, args...)
+        "_snwscanf", // _snwscanf(src, length, format, args...)
+        "_snwscanf_l", // _snwscanf_l(src, length, format, locale, args...)
+        "_fscanf_l", // _fscanf_l(src_stream, format, locale, args...)
+        "_fwscanf_l" // _fwscanf_l(src_stream, format, locale, args...)
+      ])
+  }
+
+  override predicate hasArrayWithNullTerminator(int bufParam) {
+    bufParam = [0, getFormatPosition()]
```
Comment (criemen) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll:36]: Is this true for all variations of the function?
I'm thinking especially about `fscanf` that takes a `FILE` pointer at position 0.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll [hunk 2]
```diff
@@ -0,0 +1,88 @@
...
+  }
+
+  override predicate hasArrayInput(int bufParam) { bufParam = [0, getFormatPosition()] }
```
Comment (criemen) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll:39]: same as `hasArrayWithNullTerminator`

**Review by criemen — CHANGES_REQUESTED** [review]
Some small questions, otherwise LGTM

File: cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll (hunk 1 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll:36]: Good point! I think the `0`'th parameter should only be included for the `sscanf` functions.

**Comment by MathiasVP** [comment]
Hopefully, 27b41c201648c63ee52ae7fd158679c26e124b4a addresses your comments @criemen.

**Review by criemen — DISMISSED** [review]
LGTM

**Comment by jbj** [comment]
I've added @geoffw0 as reviewer of this PR since it's now overlapping with code he's been maintaining.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    buffer = true and
+    mustWrite = true
+  }
+
+  override predicate hasSpecificReadSideEffect(ParameterIndex i, boolean buffer) {
+    buffer = true and
+    i =
+      [
+        this.(ScanfFunction).getFormatParameterIndex(),
+        this.(ScanfFunction).getFormatParameterIndex(), getLocaleParameterIndex()
```
Comment (criemen) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll:74]: `this.(ScanfFunction).getFormatParameterIndex(),` is duplicated, and should instead refer to `this.(Sscanf).getInputParameterIndex()` I guess?

**Review by criemen — CHANGES_REQUESTED** [review]
One copy-paste error, otherwise LGTM.

I'm deferring to @geoffw0 for final approval, though, as he knows the `Scanf` library.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll (hunk 3 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll:74]: Indeed. Thanks!

**Review by criemen — DISMISSED** [review]
LGTM, now waiting on @geoffw0

File: cpp/ql/src/semmle/code/cpp/commons/Scanf.qll [hunk 4]
```diff
@@ -101,6 +103,12 @@ class Snscanf extends ScanfFunction {
   override int getInputParameterIndex() { result = 0 }
 
   override int getFormatParameterIndex() { result = 2 }
+
+  /**
+   * Gets the position at which the maximum number of characters in the
+   * input string is specified.
+   */
+  int getInputLengthParameterIndex() { result = 1 }
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/commons/Scanf.qll:111]: A useful addition. :+1:

File: cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+/**
+ * The standard function `sscanf`, `fscanf` and its assorted variants
+ */
+private class SscanfModel extends ArrayFunction, TaintFunction, AliasFunction, SideEffectFunction {
+  SscanfModel() { this instanceof Sscanf or this instanceof Fscanf or this instanceof Snscanf }
+
+  override predicate hasArrayWithNullTerminator(int bufParam) {
+    bufParam = this.(ScanfFunction).getFormatParameterIndex()
+    or
+    bufParam = this.(Sscanf).getInputParameterIndex()
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll:22]: Or `(Snscanf).getInputParameterIndex()`?

Perhaps we should add a predicate to `ScanfFunction` that identifies whether the input is a string or stream so we don't need to special case this by QL class?

File: cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll [hunk 6]
```diff
@@ -0,0 +1,77 @@
...
+      else result = 2
+    )
+  }
+
+  private int getArgsStartPosition() {
+    exists(int nLength, int nLocale |
+      (if exists(getLocaleParameterIndex()) then nLocale = 1 else nLocale = 0) and
+      (if exists(getLengthParameterIndex()) then nLength = 1 else nLength = 0) and
+      result = 2 + nLocale + nLength
+    )
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll:43]: Is this not just `getNumberOfParameters()`?

File: cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll [hunk 7]
```diff
@@ -0,0 +1,77 @@
...
+  }
+
+  override predicate hasTaintFlow(FunctionInput input, FunctionOutput output) {
+    input.isParameterDeref(0) and
```
Comment (geoffw0) [inline cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll:47]:
```suggestion
    input.isParameterDeref(getInputParameterIndex()) and
```

File: cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll (hunk 5 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll:22]: Good point. Indeed, we should add a case for `Snscanf`.
> Perhaps we should add a predicate to ScanfFunction that identifies whether the input is a string or stream so we don't need to special case this by QL class?

That's a great idea! I'll prefer to just add the `Snscanf` disjunct in this PR and add a predicate to `ScanfFunction` as a follow-up PR, though.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll (hunk 6 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll:43]: You're right! Good catch.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll (hunk 7 above)
Comment (MathiasVP) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll]: You're right. It needed a cast, so I added the change in 0db54e08b8a511a3723abdef596984eaed1631eb.

File: cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll (hunk 5 above)
Comment (geoffw0) [reply cpp/ql/src/semmle/code/cpp/models/implementations/Sscanf.qll]: Sounds like a plan. :+1:

**Review by geoffw0 — APPROVED** [review]
LGTM.

**Pull request merged**